another part in the microservers frontend series (#5)
edewit/edewit.github.io: _posts/2019-12-18-microservices-frontend.md

---
layout: post
title: Microservices Frontends
date: '2019-12-18'
author: Erik Jan de Wit
---

The microservices mentioned in my [previous post][1] use the vertx event bus to communicate with each other.
An event bus is of course a great way to separate services and have them coupled loosely.
I've used the same technique for the webcomponents the carousel emits an event for the current photo that is being displayed.

```js
  this.$emit("photoId", this.photos[this.counter].id);
```

A great addition to this project would be to use the vertx client side event bus to listen to the server events and update the state of the components.

> That way then different clients like a specific photo our view will automatically reflect that.

Author: ```suggestion
That way, when different clients "perform a like" for a specific photo, our view will automatically reflect that.
```

As mentioned on the [previous post][1] I created a "stitching layer" app that uses react.
Optimally we wouldn't have to do this so we have no coupling at all, in it's simplest form we still would have a index.html that includes all the js code of the services.
But this is an easy first step that works.

> This is what the app now looks like the components marked in red are from the photo service, blue is the like service and finally green is the query service.

Author: ```suggestion
This is what the app now looks like, the components marked in red are from the photo service, blue is the like service and finally green is the query service.
```

There is also a transparent proxy to make our request go to the right microservice:
![proxy](https://lh3.googleusercontent.com/3P4IaFMKecfeGqiLfH4AuYfnpI4hqQO6PnhYRmh35XWf7gv4ySxiAM67Zk0u0TmhAcdbduVKZPGU5ZsOPAAJKyOvqYv7s5TWj1azV-jIuegxXniCsb_yKq60Szsebp_qOevRNeBzez3zHo84ujUxUyZuvu_j7hZ7n0sUoTo28ymvsm2BknxUEWlTKMegHfbffTqsvkMqlVksK6-6bxWu_ErI9UJzsFcMFoFvK51yJtMH9gorIlGFFk_K5ZqXo0eWcRnW8MdPvcGzPW9HgoTeouJJI5SGVLOjcrpmcVexUTUzjygSBMlaXyRWkfs7r7VPgiGpyw7iOQGoHa9_cFsQ_B8vggXP-slerQTFdV9c9lkrxbzq0LfRzjwR-49ctyY4G7obVWcFP_CVpxsBfW7jXOcqtDPEgJzdcR912BWnEiiFpvgnKLcl0En5hAVftD0Yj0CP6x7Ubo41T2O8bWCFXuXeSg08fubyLkqgargn_VqzDSHcU9hF4WiguHppFM1lZFc_FiOS79N3oS6PZE4W3vxqnXqrEAQtUPBhW_UORlRVILJGYXYkkB4oZyNs6n-qZp_Pu5qMWR65zgAB7Gx_dWAmnFD7jiYXBUpNAKxdf1qezrtvtzOe-dxKf0aVnfeemYh3-9AcdYx55NCKe1Rm4a4NqUwQPUkTP6vcmNqOXKUhaO7lj11nXAhCJhLR3bmk-Qe3AR5u9k32TPvW3h_twy3gzL1Pj6HG60MMSs-55tic1sw=w449-h301-no)

> In production we'll have to setup this routing some other / better way

Author: ```suggestion
In production we'll have to setup this routing some other / better way.
```

Second challenge that I bumped into was styling.
When you use webcomponents they use the shadow dom and css styling is local.
The way I fixed that now is by using css variables defined in the webcomponents and then set or overridden in the "stitching layer".
But for a production ready type application you would probably want to make your own [patternfly][2] type project or have some sort of style guide.
Downside of that is that your introducing more coupling, another way is to have shared css vars, but then you are duplicating components.

Next and final thing that we need to think about is server side rendering.
Obviously this is something that we would like to have when we have so may different smaller parts and possible different frameworks.
Server side rendering could potentially also remove the need for a "stitching layer" and when done really well help us with developing these kinds of microservices apps.

> One of the conclusions that we can make at this point is, although theoretically we could build all of these frontends in any framework, in practice we would want to minimize the number as it will increase the size of the initial download.

Author: ```suggestion
One of the conclusions that we can make at this point is, although theoretically we could build all of these frontends in any framework, in practice we would want to minimize the number, as it will increase the size of the initial download.
```

> At a minimum we would have to use the same version of the framework.

Author: ```suggestion
At a minimum we would have to use the same version of the frameworks.
```
